File: server/lib/sites.mjs
/**
 * Site-leaderboard row management: URL normalization and upsert of one
 * measurement run into the rows list (one row per site per cold/warm mode).
 */

const NAME_MAX = 80;

/**
 * Normalize whatever the submitter sent ("https://Example.com/path",
 * "www.example.com", "example.com") to a bare lowercase hostname.
 * Returns null when it cannot be read as a hostname.
 */
export function normalizeSiteUrl(input) {
  if (typeof input !== "string") return null;
  const trimmed = input.trim();
  if (!trimmed || trimmed.length > 300) return null;

  let hostname;
  try {
    hostname = new URL(/^[a-z][a-z0-9+.-]*:\/\//i.test(trimmed) ? trimmed : `https://${trimmed}`).hostname;
  } catch {
    return null;
  }

  hostname = hostname.toLowerCase().replace(/^www\./, "").replace(/\.$/, "");
  // Require a dotted hostname of sane label characters; localhost and bare
  // hosts are fine for local testing but have no place on a public board.
  if (!/^[a-z0-9]([a-z0-9-]*[a-z0-9])?(\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)+$/.test(hostname)) {
    return null;
  }
  if (hostname.length > 253) return null;
  return hostname;
}

/** "docs.example.com" -> "Example" — a display-name fallback. */
export function displayNameForUrl(url) {
  const labels = url.split(".");
  const core = labels.length >= 2 ? labels[labels.length - 2] : labels[0];
  return core.charAt(0).toUpperCase() + core.slice(1);
}

export function defaultFaviconForUrl(url) {
  return `https://icons.duckduckgo.com/ip3/${url}.ico`;
}

/**
 * Upsert one validated measurement into the rows list.
 * Existing (url, mode) row: metrics are replaced by the latest run, the test
 * counter increments, and name/favicon refresh when provided. New row: test
 * counter starts at 1.
 *
 * @returns {{rows: Array<object>, row: object, created: boolean}}
 */
export function upsertSite(rows, submission, nowIso) {
  const next = rows.map((row) => ({ ...row }));
  const index = next.findIndex(
    (row) => row.url === submission.url && row.mode === submission.mode,
  );

  const base = index === -1 ? null : next[index];
  const row = {
    name: (submission.name || base?.name || displayNameForUrl(submission.url)).slice(0, NAME_MAX),
    url: submission.url,
    favicon: submission.favicon || base?.favicon || defaultFaviconForUrl(submission.url),
    lcpRaw: submission.lcpRaw,
    lcpDelta: submission.lcpDelta,
    ttiRaw: submission.ttiRaw,
    ttiDelta: submission.ttiDelta,
    mode: submission.mode,
    tests: (base?.tests || 0) + 1,
    measuredAt: nowIso,
  };

  if (index === -1) {
    next.push(row);
  } else {
    next[index] = row;
  }
  return { rows: next, row, created: index === -1 };
}

File: backend/internal/leaderboard/sites.go
package leaderboard

import (
	"net/url"
	"regexp"
	"strings"
	"time"
)

const nameMax = 80

var (
	// A URL the submitter already gave a scheme to; anything else is assumed
	// to be a bare hostname and gets https:// prepended before parsing.
	schemePattern = regexp.MustCompile(`(?i)^[a-z][a-z0-9+.\-]*://`)

	// A dotted hostname of sane DNS labels. localhost and other bare hosts are
	// fine for local testing but have no place on a public board.
	hostnamePattern = regexp.MustCompile(`^[a-z0-9]([a-z0-9-]*[a-z0-9])?(\.[a-z0-9]([a-z0-9-]*[a-z0-9])?)+$`)
)

// NormalizeSiteURL reduces whatever the submitter sent
// ("https://Example.com/path", "www.example.com", "example.com") to a bare
// lowercase hostname. The second return is false when it cannot be read as a
// public hostname.
func NormalizeSiteURL(input string) (string, bool) {
	trimmed := strings.TrimSpace(input)
	if trimmed == "" || len(trimmed) > 300 {
		return "", false
	}

	candidate := trimmed
	if !schemePattern.MatchString(candidate) {
		candidate = "https://" + candidate
	}
	parsed, err := url.Parse(candidate)
	if err != nil {
		return "", false
	}

	hostname := strings.ToLower(parsed.Hostname())
	hostname = strings.TrimPrefix(hostname, "www.")
	hostname = strings.TrimSuffix(hostname, ".")
	if !hostnamePattern.MatchString(hostname) || len(hostname) > 253 {
		return "", false
	}
	return hostname, true
}

// DisplayNameForURL turns "docs.example.com" into "Example" — the fallback
// used when a submission carries no display name.
func DisplayNameForURL(host string) string {
	labels := strings.Split(host, ".")
	core := labels[0]
	if len(labels) >= 2 {
		core = labels[len(labels)-2]
	}
	if core == "" {
		return core
	}
	return strings.ToUpper(core[:1]) + core[1:]
}

// DefaultFaviconForURL is the favicon shown when a submission omits one.
func DefaultFaviconForURL(host string) string {
	return "https://icons.duckduckgo.com/ip3/" + host + ".ico"
}

// UpsertSite folds one validated measurement into the row that already exists
// for this (url, mode), or builds a brand new one.
//
// An existing row keeps its derived name/favicon unless the submission
// overrides them, has its metrics replaced by the latest run, and increments
// its test counter. A new row starts at one test.
func UpsertSite(existing *SiteRow, submission SiteSubmission, now time.Time) SiteRow {
	row := SiteRow{
		URL:        submission.URL,
		LCPRaw:     submission.LCPRaw,
		LCPDelta:   submission.LCPDelta,
		TTIRaw:     submission.TTIRaw,
		TTIDelta:   submission.TTIDelta,
		Mode:       submission.Mode,
		Tests:      1,
		MeasuredAt: FormatTimestamp(now),
	}

	row.Name = truncate(firstNonEmpty(submission.Name, existingName(existing), DisplayNameForURL(submission.URL)), nameMax)
	row.Favicon = firstNonEmpty(submission.Favicon, existingFavicon(existing), DefaultFaviconForURL(submission.URL))
	if existing != nil {
		row.Tests = existing.Tests + 1
	}
	return row
}

func existingName(row *SiteRow) string {
	if row == nil {
		return ""
	}
	return row.Name
}

func existingFavicon(row *SiteRow) string {
	if row == nil {
		return ""
	}
	return row.Favicon
}

func firstNonEmpty(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}

// truncate caps a string at max characters, counting runes so a multi-byte
// character can never be cut in half.
func truncate(value string, max int) string {
	runes := []rune(value)
	if len(runes) <= max {
		return value
	}
	return string(runes[:max])
}
